File: utils/src/main/java/io/github/aemogie/timble/utils/events/CancellableEvent.java
package io.github.aemogie.timble.utils.events;

import java.util.ArrayList;

public class CancellableEvent extends Event {
	private boolean cancelled = false;
	
	@Override
	protected boolean fire(ArrayList<Listener<Event>> listeners) {
		boolean success = true;
		for (Listener<Event> listener : listeners) {
			if (cancelled) break;
			success = success && (listener.fire(this) || !listener.important);
		}
		cancelled = false;
		return success;
	}
	
	public boolean cancel() {
		cancelled = true;
		return true;
	}
}

File: utils/src/main/kotlin/io/github/aemogie/timble/utils/events/CancellableEvent.kt
package io.github.aemogie.timble.utils.events

class CancellableEvent : Event() {
	private var cancelled = false
	override fun fire(listeners: ArrayList<Listener<Event?>>): Boolean {
		var success = true
		for (listener in listeners) {
			if (cancelled) break
			success = success && (listener.fire(this) || !listener.important)
		}
		cancelled = false
		return success
	}
	
	fun cancel(): Boolean {
		cancelled = true
		return true
	}
}